spam/spam_data_augmentation_tutorial.py

```diff
@@ -0,0 +1,301 @@
+# 3. **Applying Transformation Functions**: We apply a sequence of TFs to each training data point, using a random policy, to generate an augmented training set.
+# 4. **Training An End Model**: We use the augmented training set to train an LSTM model for classifying new comments as `SPAM` or `HAM`.
+
+# %% [markdown]
+# ### Data Splits in Snorkel
+#
+# We split our data into 3 sets:
+# * **Training Set**: The largest split of the dataset. These are the examples used for training, and also the ones that transformation functions are applied on.
+# * **Validation Set**: A labeled set used to tune hyperparameters and/or perform early stopping while training the classifier.
+# * **Test Set**: A labeled set for final evaluation of our classifier. This set should only be used for final evaluation, _not_ tuning.
+
+# %% [markdown]
+# ## 1. Loading Data
+
+# %% [markdown]
+# We load the Kaggle dataset and create Pandas DataFrame objects for each of the sets described above.
+# The two main columns in the DataFrames are:
+# * **`text`**: Raw text content of the comment
+# * **`label`**: Whether the comment is `SPAM` (1), `HAM` (0), or `UNKNOWN/ABSTAIN` (-1)
+#
+# For more details, check out the labeling functions [tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/spam_tutorial.ipynb).
+
+# %%
+# %load_ext autoreload
+# %autoreload 2
+
+# %%
+import os
+
+# Make sure we're running from the spam/ directory
+if os.path.basename(os.getcwd()) == "snorkel-tutorials":
+    os.chdir("spam")
+
+# %%
+from utils import load_spam_dataset
+
+df_train, _, df_valid, df_test = load_spam_dataset(delete_train_labels=False)
+
+# We pull out the label vectors for ease of use later
+Y_valid = df_valid["label"].values
+Y_train = df_train["label"].values
+Y_test = df_test["label"].values
+
+
+# %%
+df_train.head()
+
+# %% [markdown]
+# ## 2. Writing Transformation Functions
+#
+# Transformation Functions are functions that can be applied to a training example to create another valid training example. For example, for image classification problems, it is common to rotate or crop images in the training data to create new training inputs.
+#
+# Our task involves processing text. Some common ways to augment text includes replacing words with their synonyms, or replacing names entities with other entities. Applying these operations to a comment shouldn't change whether it is `SPAM` or not.
+#
+# Transformation functions in Snorkel are created with the `@transformation_function()` decorator, which wraps a function for taking a single data point and returning a transformed version of the data point.
+#
+# We start with a simple transformation function that changes a random character in the text to simulate a typo.
+
+# %%
+import string
+from snorkel.augmentation.tf import transformation_function
+
+
+@transformation_function()
+def change_character(x):
```

Comment:
Does this really help performance? Will map to a different hash bucket entirely in preprocessing

Reply:
It's hard to measure if any individual TF helps performance. But I think we should have at least one TF that doesn't rely on spacy, so we can introduce transformation functions and preprocessors separately. Also, apply() isn't currently taking that long, especially compared to the end model training.

Reply:
Removed. It wasn't helping.